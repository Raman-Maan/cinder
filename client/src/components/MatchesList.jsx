import React, { Component } from 'react';
import './styles/MatchesList.css';

import MatchesListItem from '../components/MatchesListItem';

class MatchesList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      active: false
    };
  }

  onListItemClick(id) {
    this.setState({ active: id });
    this.props.clickHandler(id);
  }

  render() {
    if (!this.props.matches || !this.props.matches.length) {
      return (
        <div className="MatchesList">
          <div className="empty">
            <h3 className="msg align-self-center">
              No matches yet. Get swiping!
            </h3>
          </div>
        </div>
      );
    }

    const listItems = this.props.matches.map(
      ({ id, title, subtitle, img, date }) => (
        <MatchesListItem
          key={id}
          title={title}
          subtitle={subtitle}
          img={img}
          date={date}
          active={id === this.state.active}
          // eslint-disable-next-line react/jsx-no-bind
          onClick={this.onListItemClick.bind(this, id)}
        />
      )
    );

    return (
      <div className="MatchesList">
        <ul>{listItems}</ul>
      </div>
    );
  }
}

export default MatchesList;
