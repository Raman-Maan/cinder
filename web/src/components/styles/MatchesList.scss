$bg-color: #ddd;
$border-color: #bbb;
$empty-msg-color: #aaa;

.MatchesList {
  width: 100%;
  height: 100%;
  max-height: 100vh;
  overflow-y: auto;

  background-color: $bg-color;
  border-right: 1px solid $border-color;

  .empty {
    display: flex;
    width: 100%;
    height: 100%;

    padding: 1em;

    justify-content: center;

    .msg {
      color: $empty-msg-color;
    }
  }

  ul {
    height: 100%;
    padding: 0;
    margin: 0;
  }
}
