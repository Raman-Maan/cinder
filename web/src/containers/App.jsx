import React, { Component } from 'react';
import { Switch, Route, withRouter } from 'react-router-dom';
import PrivateRoute from '../components/PrivateRoute';
import './styles/App.css';
import { connect } from 'react-redux';

import Layout from './Layout';
import Login from './Login';
import Signup from './Signup';

class App extends Component {
  render() {
    return (
      <div className="App">
        <Switch>
          <Route exact path="/login" component={Login} />
          <PrivateRoute
            exact
            path="/"
            authed={this.props.isAuthenticated}
            component={Layout}
          />
          <Route exact path="/signup" component={Signup} />
        </Switch>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  isAuthenticated: state.auth.isAuthenticated
});

export default withRouter(connect(mapStateToProps, null)(App));
