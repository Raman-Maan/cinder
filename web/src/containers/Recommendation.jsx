import React, { Component } from 'react';
import UserDetail from '../components/UserDetail';
import Auth from '../utils/authService';

import './styles/Recommendation.css';
export default class Recommendation extends Component {
  constructor(props) {
    super(props);
    this.fetchPotentialMatches = this.fetchPotentialMatches.bind(this);
    this.handleLike = this.handleLike.bind(this);
    this.handlePass = this.handlePass.bind(this);

    this.state = {
      potentialMatches: [],
      matchIndex: 0,
      currRecommendation: {}
    };
  }

  componentDidMount() {
    this.fetchPotentialMatches();
  }

  incrementPotentialMatchIndex() {
    this.setState(prevIndex => ({ matchIndex: prevIndex.matchIndex + 1 }));
    this.setState({
      currRecommendation: this.state.potentialMatches[this.state.matchIndex]
    });
  }

  handlePass() {
    this.submitUserAction('pass');
    this.incrementPotentialMatchIndex();
  }

  handleLike() {
    this.submitUserAction('like');
    this.incrementPotentialMatchIndex();
  }

  submitUserAction(userAction) {
    const matchedUser = this.state.potentialMatches[this.state.matchIndex].id;
    console.log(
      `Doing Action ${userAction} on ${this.state.currRecommendation.name}`
    );

    fetch(
      `/api/users/${Auth.loggedInUser.id}/matches/${matchedUser}/${userAction}`,
      { method: 'POST' }
    )
      .then(res => res.json())
      .then(res => {
        this.setState({
          // TODO
        });
      })
      .catch(err => console.error(err));
  }

  fetchPotentialMatches() {
    fetch(`/api/users/${Auth.loggedInUser.id}/recommendations`)
      .then(res => res.json())
      .then(res => {
        this.setState({
          potentialMatches: res.map(x => ({
            id: x.userID,
            name: x.userName,
            age: x.age,
            img: x.primaryPic,
            bio: x.userBio
          }))
        });
        this.setState({
          currRecommendation: this.state.potentialMatches[this.state.matchIndex]
        });
      })
      .catch(err => console.error(err));
  }

  render() {
    if (!this.state.potentialMatches.length) {
      return (
        <div className="empty">
          <h1 className="msg">There's no one new around you :(</h1>
        </div>
      );
    }

    const { img, name, age, bio } = this.state.currRecommendation;

    return (
      <UserDetail
        img={img}
        name={name}
        age={age}
        bio={bio}
        matchBtns={{
          handleLike: this.handleLike.bind(this),
          handlePass: this.handlePass.bind(this)
        }}
      />
    );
  }
}
