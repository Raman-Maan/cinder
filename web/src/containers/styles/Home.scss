.Home {
  display: flex;
  height: 100%;
  width: 100%;

  .container-fluid {
    padding: 0;
    
    .row.full {
      width: 100%;
      height: 100%;
      margin: 0;
      
      .col {
        padding: 0;
      }
    }
  }
}

.BackButton {
  margin-left: 5px;
  margin-top: 5px;
}
