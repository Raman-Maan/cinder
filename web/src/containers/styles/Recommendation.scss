$bg-color: #ddd;
$empty-msg-color: #aaa;

.empty {
  display: flex;
  width: 100%;
  height: 100%;

  background-color: $bg-color;

  padding: 1em;

  justify-content: center;
  align-items: center;

  .msg {
    color: $empty-msg-color;
  }
}
