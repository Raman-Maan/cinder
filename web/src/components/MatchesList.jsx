import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchAllMatches } from '../actions';
import PropTypes from 'prop-types';
import MatchesListItem from '../components/MatchesListItem';
import './styles/MatchesList.css';

export class MatchesList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      active: false
    };
  }

  componentDidMount() {
    this.props.fetchData(this.props.auth.userID, this.props.auth.token);
  }

  onListItemClick(id) {
    this.setState({ active: id });
    this.props.clickHandler(id);
  }

  render() {
    const msg = this.props.loading
      ? 'Loading...'
      : !this.props.matches.length
        ? 'No matches yet. Get swiping!'
        : 'There was an error loading matches';

    if (
      this.props.errored ||
      (this.props.loading && !this.props.matches.length) ||
      !this.props.matches.length
    ) {
      return (
        <div className="MatchesList">
          <div className="empty">
            <h3 className="msg align-self-center">{msg}</h3>
          </div>
        </div>
      );
    }

    const listItems = this.props.matches.map(
      ({ userID, userName, userBio, primaryPic, matchDate }) => (
        <MatchesListItem
          key={userID}
          title={userName}
          subtitle={userBio}
          img={primaryPic}
          date={matchDate}
          active={userID === this.state.active}
          // eslint-disable-next-line react/jsx-no-bind
          onClick={this.onListItemClick.bind(this, userID)}
        />
      )
    );

    return (
      <div className="MatchesList">
        <ul>{listItems}</ul>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  matches: state.match.matches,
  errored: state.match.errored,
  loading: state.match.loading,
  auth: state.auth
});

const mapDispatchToProps = dispatch => ({
  fetchData: (userID, token) => dispatch(fetchAllMatches(userID, token))
});

MatchesList.propTypes = {
  fetchData: PropTypes.func.isRequired,
  auth: PropTypes.shape({
    isFetching: PropTypes.bool,
    isAuthenticated: PropTypes.bool.isRequired,
    token: PropTypes.string,
    userID: PropTypes.string,
    message: PropTypes.string
  }),
  clickHandler: PropTypes.func.isRequired,
  matches: PropTypes.array,
  loading: PropTypes.bool,
  errored: PropTypes.bool
};

export default connect(mapStateToProps, mapDispatchToProps)(MatchesList);
