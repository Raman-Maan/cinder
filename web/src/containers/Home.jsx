import React, { Component } from 'react';
import { Button, Container, Row, Col } from 'reactstrap';

import './styles/Home.css';
import Recommendation from './Recommendation';
import MatchesList from '../components/MatchesList';
import UserDetail from '../components/UserDetail';
import Auth from '../utils/authService';

class Home extends Component {
  constructor(props) {
    super(props);

    this.fetchUserDetail = this.fetchUserDetail.bind(this);
    this.onBackButtonClick = this.onBackButtonClick.bind(this);
    this.state = {
      matches: [],
      userDetail: false,
      buttonSelected: false
    };
  }

  onBackButtonClick() {
    this.setState({ buttonSelected: true });
    this.setState({ userDetail: false });
  }

  componentDidMount() {
    this.fetchUserMatches();
  }

  fetchUserDetail(id) {
    fetch(`/api/users/${Auth.loggedInUser.id}/matches/${id}`)
      .then(res => res.json())
      .then(res => {
        this.setState({
          userDetail: {
            userID: res.userID,
            userName: res.userName,
            userPics: res.userPics,
            userAge: res.userAge,
            userGender: res.userGender,
            matchTime: new Date(),
            userBio: res.userBio ? res.userBio : 'No bio'
          }
        });
      })
      .catch(err => err);
  }

  fetchUserMatches() {
    fetch(`/api/users/${Auth.loggedInUser.id}/matches`)
      .then(res => res.json())
      .then(res => {
        this.setState({
          matches: res.map(x => ({
            id: x.userID,
            title: x.userName,
            subtitle: 'You\'ve just matched!',
            date: new Date(x.matchDate),
            img: x.primaryPic
          }))
        });
      })
      .catch(err => console.error(err));
  }

  render() {
    let rightPane = null;

    if (!this.state.userDetail) {
      rightPane = <Recommendation />;
    } else {
      rightPane = (
        <div>
          <Button
            outline
            color="primary"
            className="BackButton"
            onClick={() => this.onBackButtonClick()}
          >
            Back
          </Button>
          <UserDetail
            img={this.state.userDetail.userPics}
            name={this.state.userDetail.name}
            age={this.state.userDetail.userAge}
            bio={this.state.userDetail.userBio}
            matchDate={this.state.userDetail.matchTime}
          />
        </div>
      );
    }
    return (
      <div className="Home">
        <Container fluid>
          <Row className="full">
            <Col sm="4">
              <MatchesList
                matches={this.state.matches}
                clickHandler={this.fetchUserDetail.bind(this)}
              />
            </Col>
            <Col sm="8">{rightPane}</Col>
          </Row>
        </Container>
      </div>
    );
  }
}

export default Home;
