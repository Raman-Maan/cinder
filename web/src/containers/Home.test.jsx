import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';
import fetchMock from 'fetch-mock';

import App from './App';
import Home from './Home';
import MatchesList from '../components/MatchesList';
import Recommendation from './Recommendation';
describe('<Home />', () => {
  const testMatches = [
    {
      userName: 'Some Name 1',
      userID: 1,
      matchDate: '2018-02-03T18:09:05.000Z',
      primaryPic: null
    },
    {
      userName: 'Some Name 2',
      userID: 5,
      matchDate: '2018-01-03T18:09:05.000Z',
      primaryPic: 'http://c.min.ms/m/b/13/13022/e8250eba.jpeg'
    },
    {
      userName: 'Some Name 3',
      userID: 3,
      matchDate: '2018-02-06T18:09:05.000Z',
      primaryPic: null
    },
    {
      userName: 'Some Name 4',
      userID: 2,
      matchDate: '2018-02-04T18:09:05.000Z',
      primaryPic: 'http://c.min.ms/m/b/13/13022/e8250eba.jpeg'
    }
  ];

  beforeEach(() => {
    fetchMock.reset();
    fetchMock.restore();
    fetchMock.get(/\/api\/users\/[0-9]*\/matches/, testMatches);
    fetchMock.get(/\/api\/users\/[0-9]*\/recommendations/, []);
  });

  afterAll(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('renders without crashing', () => {
    // simple smoke test
    shallow(<Home />);
    mount(<Home />);
  });

  it('renders correctly', () => {
    const component = renderer.create(<Home />).toJSON();

    expect(component).toMatchSnapshot();
  });

  it('routes correctly', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.find(Home)).toHaveLength(1);
  });

  it('should show the <MatchesList /> component be default', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.find(MatchesList)).toHaveLength(1);
  });

  it('should show the <Recommendation /> component be default', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    );
    wrapper.setState({ userDetail: false });

    expect(wrapper.find(Recommendation)).toHaveLength(1);
  });
});
