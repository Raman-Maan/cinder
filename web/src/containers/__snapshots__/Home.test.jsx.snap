// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<div
  className="Home"
>
  <div
    className="container-fluid"
  >
    <div
      className="full row"
    >
      <div
        className="col col-sm-4"
      >
        <div
          className="MatchesList"
        >
          <div
            className="empty"
          >
            <h3
              className="msg align-self-center"
            >
              No matches yet. Get swiping!
            </h3>
          </div>
        </div>
      </div>
      <div
        className="col col-sm-8"
      >
        <div
          className="empty"
        >
          <h1
            className="msg"
          >
            There's no one new around you :(
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;
